lib/oauth/apple/idtoken.go

```diff
@@ -0,0 +1,151 @@
+package apple
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/go-jose/go-jose/v3"
+	"github.com/go-jose/go-jose/v3/jwt"
+	"golang.org/x/oauth2"
+)
+
+type TokenReader struct {
+	keyStore rsaKeyStore
+	clientID string
+}
+
+func NewTokenReader(clientID string) *TokenReader {
+	return &TokenReader{
+		clientID: clientID,
+		keyStore: newSigningKeyStore(&http.Client{
+			Timeout: 10 * time.Second,
+		}),
+	}
+}
+
+func (tr *TokenReader) GetUserDetailsFromToken(t *oauth2.Token) (string, string, error) {
+	claims, err := tr.getAppleIDTokenClaims(t)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to get apple ID token claims: %w", err)
+	}
+
+	return extractUserFromClaims(claims, tr.clientID)
+}
+
+func (tr *TokenReader) getAppleIDTokenClaims(t *oauth2.Token) (map[string]interface{}, error) {
+	idTokenVal := t.Extra("id_token")
+	if idTokenVal == nil {
+		return nil, fmt.Errorf("missing id_token")
+	}
+
+	idToken, ok := idTokenVal.(string)
+
+	if !ok {
+		return nil, fmt.Errorf("id_token is not a string")
+	}
+
+	parsedIDToken, err := jwt.ParseSigned(idToken)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var hdr *jose.Header
+
+	for i := range parsedIDToken.Headers {
+		th := parsedIDToken.Headers[i]
+		if th.Algorithm == "RS256" {
+			hdr = &th
+			break
+		}
+	}
+
+	if hdr == nil {
+		return nil, fmt.Errorf("no RS256 key header found")
+	}
+
+	appleRSA, err := tr.keyStore.get(hdr.KeyID)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get apple RSA key: %w", err)
+	}
+
+	claims := make(map[string]interface{})
+	err = parsedIDToken.Claims(appleRSA, &claims)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to verify claims: %w", err)
+	}
+
+	return claims, nil
+}
+
+func extractUserFromClaims(claims map[string]interface{}, clientID string) (string, string, error) {
```

Comment:
can we use https://pkg.go.dev/github.com/go-jose/go-jose/v3/jwt#Claims.Validate for the standard `iss`, `aud`, and `exp` claims? then our code can focus on our unique values like `sub` and `email`.

Reply:
Will see what I can do here.  The problem is if we read into `jwt.Claims` we lose visibility into `sub` and `email` - but there might be something we can do with a custom `Claims` type in the apple package that embeds / implements its own json unmarshaler that lets us leverage it.

Reply:
i'd normally verify claims like `iss`, `aud`, and `exp` during the decoding step. is it easier to get at `sub` and `email` if you move the others?

Reply:
I kinda like how this turned out let me know what you think 95e1df396fb52c19f1f395376896febb5f12d028